src/login/login.component.ts

```diff
@@ -1,81 +1,50 @@
-import {AlertService, AuthenticationService, TokenService} from '../_services';
-import {User} from '../_models';
+import {User} from '../_models/User';
 import {sha512} from 'js-sha512';
-import {FormBuilder} from '@angular/forms';
+import {FormGroup, FormControl, Validators} from '@angular/forms';
+import {AlertService} from '../_services/alert.service';
+import {AuthenticationService} from '../_services/authentication.service';
+import {TokenService} from '../_services/token.service';
 
 @Component({
   selector: 'app-login',
   templateUrl: './login.component.html',
   styleUrls: ['./login.component.css']
 })
-export class LoginComponent implements OnInit {
-
-  objectValues = Object.values;
-  model: User = {};
+export class LoginComponent {
+  form: FormGroup = new FormGroup({
+    username: new FormControl(
+      null, [Validators.required]),
+    password: new FormControl(
+      null, [Validators.required])
+  });
   loading = false;
-  returnUnion: string;
-  unions = {
-    mydata: {key: 'mydata', name: 'My Data', url: 'https://mydata.webtree.org/applyToken'},
-    imprint: {key: 'imprint', name: 'Imprint', url: 'https://imprint.webtree.org/applyToken'}
-  };
-  loggedIn = false;
-
-  @ViewChild('redirectForm', {read: ElementRef, static: true}) redirectForm: ElementRef;
 
   constructor(private route: ActivatedRoute,
               private router: Router,
-              private formBuilder: FormBuilder,
               private authenticationService: AuthenticationService,
               private tokenService: TokenService,
               private alertService: AlertService) {
   }
 
-  async ngOnInit() {
-    this.returnUnion = this.route.snapshot.queryParams.returnUnion;
-    this.loggedIn = await this.authenticationService.isAuthorized();
-    this.redirectIfNeeded();
-  }
-
-  login() {
+  onSubmit({username, password}) {
     this.loading = true;
-    const user: User = {username: this.model.username, password: sha512(this.model.password)};
-    this.authenticationService.login(user)
-      .subscribe(
-        res => {
-          this.tokenService.saveToken(JSON.parse(res).token);
-          this.alertService.success('Logged in successfully');
-          this.redirectIfNeeded();
-        },
-        error => {
+    const user: User = {username, password: sha512(password)};
+    return this.authenticationService.login(user)
+      .subscribe(token => {
           this.loading = false;
-          console.log(error);
-          if (error.status === 401) {
-            this.alertService.error(error.error);
-          } else {
-            throw  error;
+          if (token === null) {
+            this.alertService.error('Invalid username or password');
+            return;
+          }
+
+          this.tokenService.saveToken(token);
+          this.alertService.success('Logged in successfully');
+          if (!this.authenticationService.redirectToUnionIfNeeded(this.route.snapshot)) {
```

Comment:
Can we have something like this:

```suggestion
          this.authenticationService.redirectToUnionIfNeeded(this.route.snapshot))
            .orElse(() -> this.router.navigate(['/select-union']))
```

Reply:
Or with callback in parameter at least. :)